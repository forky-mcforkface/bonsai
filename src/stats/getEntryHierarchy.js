/*
 * @flow
 */

import type { ChunkID, RawStats } from '../types/Stats';

import getChunkName from './getChunkName';
import getChunkNamesFromImportedModules from './getChunkNamesFromImportedModules';
import getEntryChunks from './getEntryChunks';

export type Child = {
  id: ChunkID,
  ids: Array<ChunkID>,
  name: string,
  names: Array<string>,
  children: Array<Child>,
};

export type FlatChunk = {
  id: ChunkID,
  name: string,
  indent: number,
};

function getChildrenForChunk(
  stats,
  importedChunkNames,
  parentChunk,
  existingParents: Array<ChunkID>,
): Array<Child> {
  return getEntryChunks(stats)
    .filter((chunk) =>
      chunk.parents.includes(parentChunk.id) &&
      !existingParents.includes(chunk.id)
    )
    .map((chunk) => ({
      id: chunk.id,
      ids: [chunk.id],
      name: getChunkName(chunk, importedChunkNames),
      names: chunk.names,
      children: getChildrenForChunk(
        stats,
        importedChunkNames,
        chunk,
        existingParents.concat(chunk.id),
      ),
    }));
}

export const ROOT_ID = Number.MIN_SAFE_INTEGER;

export default function getEntryHierarchy(
  stats: RawStats,
): Child {
  const importedChunkNames = getChunkNamesFromImportedModules(stats);

  return {
    id: ROOT_ID,
    ids: [],
    name: '',
    names: [],
    children: getEntryChunks(stats)
      .filter((chunk) => chunk.parents.length === 0)
      .map((chunk) => ({
        id: chunk.id,
        ids: [chunk.id],
        name: getChunkName(chunk, importedChunkNames),
        names: chunk.names,
        children: getChildrenForChunk(
          stats,
          importedChunkNames,
          chunk,
          [chunk.id],
        ),
      })),
  };
}

export function flattenChunksByParent(
  chunksByParent: Array<Child>,
): Array<FlatChunk> {
  const flatChunks = [];
  const visitedChunks = {};

  function appendChildren(children: Array<Child>, indent: number) {
    children.forEach((child) => {
      if (!visitedChunks[child.id]) {
        flatChunks.push({
          id: child.id,
          name: child.name,
          indent: indent,
        });
        visitedChunks[child.id] = true;
      }

      appendChildren(child.children, indent + 1);
    });
  }

  appendChildren(chunksByParent, 0);

  return flatChunks;
}
