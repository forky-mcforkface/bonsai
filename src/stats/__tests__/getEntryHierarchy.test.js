/*
 * @flow
 */

import {defaultChunk} from '../../__test_helpers__/defaults';
import getEntryHierarchy, {
  flattenChunksByParent,
} from '../getEntryHierarchy';

const stats = {
  chunks: [
    defaultChunk({
      id: 0,
      names: ['chunk-zero'],
    }),
    defaultChunk({
      id: 1,
      names: ['chunk-one'],
      parents: [0],
    }),
    defaultChunk({
      id: 2,
      names: ['chunk-two'],
    }),
  ],
  modules: [],
};

describe('getEntryHierarchy', () => {
  it('should list all the chunks', () => {
    const result = getEntryHierarchy(stats);

    expect(result).toMatchSnapshot();
  });

  it('should flatten the list of all chunks', () => {
    const result = flattenChunksByParent(getEntryHierarchy(stats).children);

    expect(result).toEqual([
      {id: 0, name: 'chunk-zero', indent: 0},
      {id: 1, name: 'chunk-one', indent: 1},
      {id: 2, name: 'chunk-two', indent: 0},
    ]);
  });
});
