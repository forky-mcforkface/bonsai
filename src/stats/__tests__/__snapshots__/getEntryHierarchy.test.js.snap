// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEntryHierarchy should list all the chunks 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "id": 1,
          "ids": Array [
            1,
          ],
          "name": "chunk-one",
          "names": Array [
            "chunk-one",
          ],
        },
      ],
      "id": 0,
      "ids": Array [
        0,
      ],
      "name": "chunk-zero",
      "names": Array [
        "chunk-zero",
      ],
    },
    Object {
      "children": Array [],
      "id": 2,
      "ids": Array [
        2,
      ],
      "name": "chunk-two",
      "names": Array [
        "chunk-two",
      ],
    },
  ],
  "id": -9007199254740991,
  "ids": Array [],
  "name": "",
  "names": Array [],
}
`;
